README.md
# video2dataset
[![pypi](https://img.shields.io/pypi/v/video2dataset.svg)](https://pypi.python.org/pypi/video2dataset)

Easily create large video dataset from video urls

![Video2dataset design overview](video2dataset_overview.png)]

Checkout the [design doc](https://docs.google.com/document/d/1_TD2KQLkEegszq4Eip568fc6cWnh9h0Jqj4Lc88t9Y0/edit#)

## Install

pip install video2dataset

## Examples


## Usage

First get some video url list. For example:
```
echo 'https://www.youtube.com/watch?v=0WfKzVqdQqo' >> myvidlist.txt
```

Then, run the tool:

```
video2dataset --url_list=myvidlist.txt --output_folder=output_folder
```

The tool will then automatically download the urls and store them with that format:
* output_folder
    * 00000
        * 000000000.mp4
        * 000000001.mp4
        * 000000002.mp4

or as this format if choosing webdataset:
* output_folder
    * 00000.tar containing:
        * 000000000.mp4
        * 000000001.mp4
        * 000000002.mp4

with each number being the position in the list. The subfolders avoids having too many files in a single folder.

If **captions** are provided, they will be saved as 0.txt, 1.txt, ...

This can then easily be fed into machine learning training or any other use case.

Also .json files named 0.json, 1.json,... are saved with these keys:
* url
* caption
* key of the form 000010005 : the first 5 digits are the shard id, the last 4 are the index in the shard
* status : whether the download succeeded
* error_message

Also a .parquet file will be saved with the same name as the subfolder/tar files containing these same metadata.
It can be used to analyze the results efficiently.

.json files will also be saved with the same name suffixed by _stats, they contain stats collected during downloading (download time, number of success, ...)

## Python examples

* [Here's an example PyTorch data loader for reading the output of video2dataset](https://github.com/iejMac/video2dataset/blob/main/examples/data_loader.py)


## API

This module exposes a single function `download` which takes the same arguments as the command line tool:

* **url_list** A file with the list of url of images to download. It can be a folder of such files. (*required*)
* **output_folder** The path to the output folder. (default *"images"*)
* **processes_count** The number of processes used for downloading the pictures. This is important to be high for performance. (default *1*)
* **encode_format** encode format (default *mp4*)
* **output_format** decides how to save pictures (default *files*)
  * **files** saves as a set of subfolder containing pictures
  * **webdataset** saves as tars containing pictures
  * **parquet** saves as parquet containing pictures as bytes
  * **tfrecord** saves as tfrecord containing pictures as bytes
  * **dummy** does not save. Useful for benchmarks
* **input_format** decides how to load the urls (default *txt*)
  * **txt** loads the urls as a text file of url, one per line
  * **csv** loads the urls and optional caption as a csv
  * **tsv** loads the urls and optional caption as a tsv
  * **tsv.gz** loads the urls and optional caption as a compressed (gzip) tsv.gz
  * **json** loads the urls and optional caption as a json
  * **parquet** loads the urls and optional caption as a parquet
* **url_col** the name of the url column for parquet and csv (default *url*)
* **caption_col** the name of the caption column for parquet and csv (default *None*)
* **clip_col** the name of the column with a list of timespans for each clip (defualt *None*)
* **save_additional_columns** list of additional columns to take from the csv/parquet files and save in metadata files (default *None*)
* **number_sample_per_shard** the number of sample that will be downloaded in one shard (default *10000*)
* **timeout** maximum time (in seconds) to wait when trying to download an image (default *10*)
* **video_size** size of video frames (default *360*)
* **resize_mode** what resizing transformations to apply to video resolution (default *None*)
  * **scale** scale video keeping aspect ratios (currently always picks video height)
  * **crop** center crop to video_size x video_size
  * **pad** center pad to video_size x video_size
* **video_fps** what FPS to resample the video to. If < 0 then video FPS remains unchanged (default *-1*)
* **enable_wandb** whether to enable wandb logging (default *False*)
* **wandb_project** name of W&B project used (default *video2dataset*)
* **oom_shard_count** the order of magnitude of the number of shards, used only to decide what zero padding to use to name the shard files (default *5*)
* **distributor** choose how to distribute the downloading (default *multiprocessing*)
  * **multiprocessing** use a multiprocessing pool to spawn processes
  * **pyspark** use a pyspark session to create workers on a spark cluster (see details below)
* **subjob_size** the number of shards to download in each subjob supporting it, a subjob can be a pyspark job for example (default *1000*)
* **incremental_mode** Can be "incremental" or "overwrite". For "incremental", video2dataset will download all the shards that were not downloaded, for "overwrite" video2dataset will delete recursively the output folder then start from zero (default *incremental*)
* **tmp_dir** name of temporary directory in your file system (default */tmp*)
* **yt_metadata_args** dict of YouTube metadata arguments (default *None*, more info below)


### Download YouTube metadata & subtitles:
#### Usage

**Note.** Requires webvtt installed for subtitles formatting: `pip install webvtt-py`.

```py
if __name__ == '__main__':

    yt_metadata_args = {
        'writesubtitles': True, # whether to write subtitles to a file
        'subtitleslangs': ['en'], # languages of subtitles (right now support only one language)
        'writeautomaticsub': True, # whether to write automatic subtitles
        'get_info': True # whether to save a video meta data into the output JSON file
    }

    video2dataset(
        url_list='test.parquet',
        input_format='parquet',
        output_format='files',
        output_folder='audio',
        yt_metadata_args=yt_metadata_args
    )
```

#### Output
For every sample save the info dict into an output file:

```json
{
    "url": "https://www.youtube.com/watch?v=q2ZOEFAaaI0",
    "key": "000000000",
    "status": "success",
    "error_message": null,
    "info": {
        "id": "q2ZOEFAaaI0",
        "title": "Q-learning with numpy and OpenAI Taxi-v2 \ud83d\ude95 (tutorial)",
        "thumbnail": "https://i.ytimg.com/vi_webp/q2ZOEFAaaI0/maxresdefault.webp",
        "description": "We'll train an Q-learning agent with Numpy that learns to play Taxi-v2. Where he must take a passenger at one location and drop him off at another as fast as possible. \ud83d\ude95\n\nThis video is part of the Deep Reinforcement Learning course with tensorflow \ud83d\udd79\ufe0f a free series of blog posts and videos \ud83c\udd95 about Deep Reinforcement Learning, where we'll learn the main algorithms, and how to implement them with Tensorflow : https://simoninithomas.github.io/Deep_reinforcement_learning_Course/\n\nIf you're new in Reinforcement Learning, please read first my article \"An introduction to Reinforcement Learning\": https://medium.freecodecamp.org/an-introduction-to-reinforcement-learning-4339519de419\n\nThe Q-learning article: https://medium.freecodecamp.org/diving-deeper-into-reinforcement-learning-with-q-learning-c18d0db58efe\n\nThe Q-learning notebook: https://github.com/simoninithomas/Deep_reinforcement_learning_Course/blob/master/Q%20learning/Taxi-v2/Q%20Learning%20with%20OpenAI%20Taxi-v2%20video%20version.ipynb\n\nThis is my first video so if you have some feedbacks and advice please comment below.\n\nMoreover if you have some questions you can ask me in the comments.\n\nDon't forget to subscribe ! And to follow me on social media:\nTwitter: https://twitter.com/ThomasSimonini\nFacebook: https://www.facebook.com/thomas.simonini.3\n\nKeep learning, stay awesome!",
        "uploader": "Thomas Simonini",
        "uploader_id": "UC8XuSf1eD9AF8x8J19ha5og",
        "uploader_url": "http://www.youtube.com/channel/UC8XuSf1eD9AF8x8J19ha5og",
        "channel_id": "UC8XuSf1eD9AF8x8J19ha5og",
        "channel_url": "https://www.youtube.com/channel/UC8XuSf1eD9AF8x8J19ha5og",
        "duration": 776,
        "view_count": 44988,
        "average_rating": null,
        "age_limit": 0,
        "webpage_url": "https://www.youtube.com/watch?v=q2ZOEFAaaI0",
        "categories": [
            "Education"
        ],
        "tags": [
            "AI",
            "Q-learning",
            "tutorial",
            "numpy",
            "reinforcement-learning",
            "programming",
            "openai",
            "deep-learning",
            "Machine-learning"
        ],
        "playable_in_embed": true,
        "live_status": null,
        "release_timestamp": null,
        "comment_count": 126,
        "chapters": [
            ...
        ],
       ...
    }
}
```

And subtitles:
```json
"subtitles": [
            {
                "start": "00:00:02.389",
                "end": "00:00:02.399",
                "lines": [
                    "hello and welcome if you want to study"
                ]
            },
            {
                "start": "00:00:04.280",
                "end": "00:00:04.290",
                "lines": [
                    "different phasma learning don't go"
                ]
            },
            ...
]
```


## Incremental mode

If a first download got interrupted for any reason, you can run again with --incremental "incremental" (this is the default) and using the same output folder , the same number_sample_per_shard and the same input urls, and video2dataset will complete the download.

## Output format choice

video2dataset support several formats. There are trade off for which to choose:
* files: this is the simplest one, images are simply saved as files. It's good for up to 1M samples on a local file system. Beyond that performance issues appear very fast. Handling more than a million files in standard filesystem does not work well.
* webdataset: webdataset format saves samples in tar files, thanks to [webdataset](https://webdataset.github.io/webdataset/) library, this makes it possible to load the resulting dataset fast in both pytorch, tensorflow and jax. Choose this for most use cases. It works well for any filesystem
* parquet: parquet is a columnar format that allows fast filtering. It's particularly easy to read it using pyarrow and pyspark. Choose this if the rest of your data ecosystem is based on pyspark. [petastorm](https://github.com/uber/petastorm) can be used to read the data but it's not as easy to use as webdataset
* tfrecord: tfrecord is a protobuf based format. It's particularly easy to use from tensorflow and using [tf data](https://www.tensorflow.org/guide/data). Use this if you plan to use the dataset only in the tensorflow ecosystem. The tensorflow writer does not use fsspec and as a consequence supports only a limited amount of filesystem, including local, hdfs, s3 and gcs. It is also less efficient than the webdataset writer when writing to other filesystems than local, losing some 30% performance.

## File system support

Thanks to [fsspec](https://filesystem-spec.readthedocs.io/en/latest/), video2dataset supports reading and writing files in [many file systems](https://github.com/fsspec/filesystem_spec/blob/6233f315548b512ec379323f762b70764efeb92c/fsspec/registry.py#L87).
To use it, simply use the prefix of your filesystem before the path. For example `hdfs://`, `s3://`, `http://`, or `gcs://`.
Some of these file systems require installing an additional package (for example s3fs for s3, gcsfs for gcs).
See fsspec doc for all the details.

If you need specific configuration for your filesystem, you may handle this problem by using the [fsspec configuration system](https://filesystem-spec.readthedocs.io/en/latest/features.html#configuration) that makes it possible to create a file such as `.config/fsspec/s3.json` and have information in it such as:
```
{
  "s3": {
    "client_kwargs": {
            "endpoint_url": "https://some_endpoint",
            "aws_access_key_id": "your_user",
           "aws_secret_access_key": "your_password"
    }
  }
}
```
Which may be necessary if using s3 compatible file systems such as [minio](https://min.io/). That kind of configuration also work for all other fsspec-supported file systems.

## Distribution modes

video2dataset supports several distributors.
* multiprocessing which spawns a process pool and use these local processes for downloading
* pyspark which spawns workers in a spark pool to do the downloading

multiprocessing is a good option for downloading on one machine, and as such it is the default.
Pyspark lets video2dataset use many nodes, which makes it as fast as the number of machines.
It can be particularly useful if downloading datasets with more than a billion image.

### pyspark configuration

In order to use video2dataset with pyspark, you will need to do this:
1. `pip install pyspark`
2. use the `--distributor pyspark` option
3. tweak the `--subjob_size 1000` option: this is the number of images to download in each subjob. Increasing it will mean a longer time of preparation to put the feather files in the temporary dir, a shorter time will mean sending less shards at a time to the pyspark job.

By default a local spark session will be created.
You may want to create a custom spark session depending on your specific spark cluster.


## For development

Either locally, or in [gitpod](https://gitpod.io/#https://github.com/iejMac/video2dataset) (do `export PIP_USER=false` there)

Setup a virtualenv:

```
python3 -m venv .env
source .env/bin/activate
pip install -e .
```

to run tests:
```
pip install -r requirements-test.txt
```
then 
```
make lint
make test
```

You can use `make black` to reformat the code

`python -m pytest -x -s -v tests -k "dummy"` to run a specific test

## Benchmarks

## Special Contributors:

* [ChatGPT](https://chat.openai.com) - FrameSubsampler implementation


HISTORY.md
## 1.0.0

* it works a bit


video2dataset/main.py
"""Create dataset from video links and metadata."""


from typing import List, Optional
import fire

from .logger import LoggerProcess
from .data_writer import (
    WebDatasetSampleWriter,
    FilesSampleWriter,
    ParquetSampleWriter,
    TFRecordSampleWriter,
    DummySampleWriter,
)
from .input_sharder import InputSharder
from .distributor import multiprocessing_distributor, pyspark_distributor
import fsspec
import sys
import signal
import os
from .worker import Worker


def video2dataset(
    url_list: str,
    output_folder: str = "videos",
    processes_count: int = 1,
    thread_count: int = 16,
    output_format: str = "files",
    input_format: str = "txt",
    url_col: str = "url",
    caption_col: Optional[str] = None,
    clip_col: Optional[str] = None,
    save_additional_columns: Optional[List[str]] = None,
    number_sample_per_shard: int = 10000,
    enable_wandb: bool = False,
    wandb_project: str = "video2dataset",
    oom_shard_count: int = 5,
    distributor: str = "multiprocessing",
    subjob_size: int = 1000,
    incremental_mode: str = "incremental",
    max_shard_retry: int = 1,
    video_size: int = 360,
    video_fps: int = -1,
    resize_mode: Optional[List[str]] = None,
    timeout: int = 60,
    tmp_dir: str = "/tmp",
    yt_metadata_args: dict = None,
    encode_formats: dict = None,
):
    """
    create video dataset from video links
    """

    config_parameters = dict(locals())

    if encode_formats is None:
        encode_formats = {"video": "mp4"}

    def make_path_absolute(path):
        fs, p = fsspec.core.url_to_fs(path)
        if fs.protocol == "file":
            return os.path.abspath(p)
        return path

    output_folder = make_path_absolute(output_folder)
    url_list = make_path_absolute(url_list)

    logger_process = LoggerProcess(output_folder, enable_wandb, wandb_project, config_parameters)

    tmp_path = output_folder + "/_tmp"
    fs, tmp_dir = fsspec.core.url_to_fs(tmp_path)
    if not fs.exists(tmp_dir):
        fs.mkdir(tmp_dir)

    def signal_handler(signal_arg, frame):  # pylint: disable=unused-argument
        try:
            fs.rm(tmp_dir, recursive=True)
        except Exception as _:  # pylint: disable=broad-except
            pass
        logger_process.terminate()
        sys.exit(0)

    signal.signal(signal.SIGINT, signal_handler)

    save_caption = caption_col is not None

    fs, output_path = fsspec.core.url_to_fs(output_folder)

    if not fs.exists(output_path):
        fs.mkdir(output_path)
        done_shards = set()
    else:
        if incremental_mode == "incremental":
            done_shards = set(int(x.split("/")[-1].split("_")[0]) for x in fs.glob(output_path + "/*.json"))
        elif incremental_mode == "overwrite":
            fs.rm(output_path, recursive=True)
            fs.mkdir(output_path)
            done_shards = set()
        else:
            raise ValueError(f"Unknown incremental mode {incremental_mode}")

    logger_process.done_shards = done_shards
    logger_process.start()

    input_sharder = InputSharder(
        url_list,
        input_format,
        url_col,
        caption_col,
        clip_col,
        save_additional_columns,
        number_sample_per_shard,
        done_shards,
        tmp_path,
    )

    if output_format == "webdataset":
        sample_writer_class = WebDatasetSampleWriter
    elif output_format == "parquet":
        sample_writer_class = ParquetSampleWriter  # type: ignore
    elif output_format == "files":
        sample_writer_class = FilesSampleWriter  # type: ignore
    elif output_format == "tfrecord":
        sample_writer_class = TFRecordSampleWriter  # type: ignore
    elif output_format == "dummy":
        sample_writer_class = DummySampleWriter  # type: ignore
    else:
        raise ValueError(f"Invalid output format {output_format}")

    worker = Worker(
        sample_writer_class=sample_writer_class,
        save_caption=save_caption,
        output_folder=output_folder,
        column_list=input_sharder.column_list,
        thread_count=thread_count,
        timeout=timeout,
        number_sample_per_shard=number_sample_per_shard,
        oom_shard_count=oom_shard_count,
        video_size=video_size,
        resize_mode=resize_mode,
        video_fps=video_fps,
        tmp_dir=tmp_dir,
        yt_metadata_args=yt_metadata_args,
        encode_formats=encode_formats,
    )

    print("Starting the downloading of this file")
    if distributor == "multiprocessing":
        distributor_fn = multiprocessing_distributor
    elif distributor == "pyspark":
        distributor_fn = pyspark_distributor
    else:
        raise ValueError(f"Distributor {distributor} not supported")

    distributor_fn(
        processes_count,
        worker,
        input_sharder,
        subjob_size,
        max_shard_retry,
    )
    logger_process.join()
    fs.rm(tmp_dir, recursive=True)


def main():
    fire.Fire(video2dataset)


if __name__ == "__main__":
    main()
